src/v/cloud_storage/segment_meta_cstore.h

```diff
@@ -0,0 +1,848 @@
+        std::advance(it, index - curr_ix);
+        return it;
+    }
+
+    size_t size() const { return _size; }
+
+    size_t mem_use() const {
+        if (!_tail.has_value()) {
+            return sizeof(*this);
+        }
+        return sizeof(*this) + _tail.value().mem_use();
+    }
+
+    const_iterator begin() const {
+        if (_tail.has_value()) {
+            decoder_t decoder(
+              _tail->get_initial_value(),
+              _tail->get_row_count(),
+              _tail->share(),
+              _delta_alg);
+            return const_iterator(std::move(decoder), _head, _size);
+        } else if (_size != 0) {
+            // special case, data is only stored in the buffer
+            // not in the compressed column
+            decoder_t decoder(0, 0, iobuf(), _delta_alg);
+            return const_iterator(std::move(decoder), _head, _size);
+        }
+        return end();
+    }
+
+    const_iterator end() const { return const_iterator(); }
+
+    const_iterator find(value_t value) const {
+        return pred_search<std::equal_to<value_t>>(value);
+    }
+
+    const_iterator upper_bound(value_t value) const {
+        return pred_search<std::greater<value_t>>(value);
+    }
+
+    const_iterator lower_bound(value_t value) const {
+        return pred_search<std::greater_equal<value_t>>(value);
+    }
+
+    std::optional<value_t> last_value() const {
+        if (_size == 0) {
+            return std::nullopt;
+        }
+        auto ix = (_size - 1) & index_mask;
+        return _head.at(ix);
+    }
+
+    bool contains(value_t value) const { return find(value) != end(); }
+
+    void prefix_truncate(value_t new_start) {
+        segment_meta_column_frame<value_t, delta_t> tmp(_delta_alg);
+        for (auto it = find(new_start); it != end(); ++it) {
+            tmp.append(*it);
+        }
+        _head = tmp._head;
+        _tail = std::move(tmp._tail);
+        _size = tmp._size;
+        _last_row = tmp._last_row;
+    }
+
+    void prefix_truncate_ix(uint32_t ix) {
+        segment_meta_column_frame<value_t, delta_t> tmp(_delta_alg);
+        for (auto it = at(ix); it != end(); ++it) {
+            tmp.append(*it);
+        }
+        _head = tmp._head;
+        _tail = std::move(tmp._tail);
+        _size = tmp._size;
+        _last_row = tmp._last_row;
+    }
+
+private:
+    template<class PredT>
+    const_iterator pred_search(value_t value) const {
+        PredT pred;
+        for (auto it = begin(); it != end(); ++it) {
+            if (pred(*it, value)) {
+                return it;
+            }
+        }
+        return end();
+    }
+
+    std::array<value_t, buffer_depth> _head{};
+    mutable std::optional<encoder_t> _tail{std::nullopt};
+    const delta_t _delta_alg;
+    size_t _size{0};
+    std::optional<hint_t> _last_row;
+};
+
+/// Column iterator
+///
+/// The core idea here is that the iterator stores an immutable
```

Comment:
The lifetime of underlying column iobufs is a potential worry here: where do we expect iterators to be used, would they exist anywhere long-lived that could cause excess memory use by keeping the iobufs alive?

Reply:
Currently, the iterators are not cached anywhere. And even if the whole column-store will be regenerated during compaction step, the iterator can only hold iobufs which belong to regenerated frames.